freezer/src/freezer_files.rs

```diff
@@ -0,0 +1,366 @@
+use crate::internal_error;
+use ckb_error::Error;
+use std::collections::BTreeMap;
+use std::convert::TryInto;
+use std::fs::{self, File};
+use std::io::{Read, Write};
+use std::io::{Seek, SeekFrom};
+use std::path::{Path, PathBuf};
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+
+const MAX_FILE_SIZE: u64 = 2 * 1_000 * 1_000 * 1_000;
+const INDEX_FILE_NAME: &str = "INDEX";
+const INDEX_ENTRY_SIZE: u64 = 12;
+
+pub type FileId = u32;
+
+pub(crate) struct Head {
+    pub(crate) file: File,
+    pub(crate) bytes: u64,
+}
+
+impl Head {
+    pub fn new(file: File, bytes: u64) -> Self {
+        Head { file, bytes }
+    }
+
+    pub fn write(&mut self, data: &[u8]) -> Result<(), Error> {
+        self.file.write_all(data).map_err(internal_error)?;
+        self.bytes = self.bytes + data.len() as u64;
+        Ok(())
+    }
+}
+
+pub struct FreezerFiles {
+    pub files: BTreeMap<FileId, File>,
+    pub(crate) tail_id: FileId,
```

Comment:
`tail_id` is used only once, and it can be retrieved based on `index`. Hence I think it is unnecessary to store as a field.